### PR TITLE
web-ui: replace textarea tag with codemirror.

### PR DESCRIPTION
**Pull request by pjmp**
Hey if you guys are interested, I've added a web text editor to make viewing bearable, please note that I'm no html/css expert.

It looks like this:

image

Technical details:
- Replace textarea with [codemirror v5](https://codemirror.net/5/)
- Minor style changes to accommodate the editor

If this doesn't align with your goal feel free to close this pr.

Thank you for your amazing work.

### CODE REVIEW

File: bin/web-ui/index.html [hunk 1]
```diff
@@ -48,7 +62,8 @@
           <label for="code">OCaml code</label>
           <button type="button" id="format">format</button>
         </div>
-        <textarea name="code" id="code" rows="50" cols="120" spellcheck="false"></textarea>
+        <textarea name="code" id="code" class="language-mllike" spellcheck="false"></textarea>
+        </textarea>
```
Comment (Khady) [inline bin/web-ui/index.html:66]:
```suggestion
```

duplicated `</textarea>`

File: bin/web-ui/main.ml [hunk 2]
```diff
@@ -425,6 +425,11 @@ let onload _event =
             match code_formatted with
             | Ok code_formatted ->
                 code_input##.value := Js.string code_formatted
+                (* somehow, we have to replace above line with
+                let formatted = Js.string code_formatted in
+                (* editor is attached to window object *)
+                editor.doc.setValue(formatted)
+                *)
```
Comment (Khady) [inline bin/web-ui/main.ml:432]: I don't understand this comment. Is there something to do?

**Comment by Khady** [comment]
Thanks, it looks much better with an editor indeed.

@davesnx do you have an opinion on this one?

File: bin/web-ui/main.ml (hunk 2 above)
Comment (pjmp) [reply bin/web-ui/main.ml:432]: @Khady thanks for looking into, I put out this just to gauge interest.
I don't have much experience using ocaml and I'm self learning. I saw your announcement in [discuss thread](https://discuss.ocaml.org/t/ann-ocamlformat-web-configurator/10103) and it's a really useful tool.

With that said, I haven't used `js_of_ocaml` so i dont know how to call `editor.doc.setValue` in the window object which is needed to update the editor with newly formatted string. That's why i left that comment.

Side note: I couldn't build the project so if there's an instruction how to build the project, I'll be happy to fix this.

File: bin/web-ui/index.html (hunk 1 above)
Comment (pjmp) [reply bin/web-ui/index.html]: removed

File: bin/web-ui/main.ml (hunk 2 above)
Comment (Khady) [reply bin/web-ui/main.ml]: ok I see the problem now. But I don't know the immediate solution ^^' Will try to have a look

> Side note: I couldn't build the project so if there's an instruction how to build the project, I'll be happy to fix this.

That is a good remark, I should add a readme

Comment (pjmp) [reply bin/web-ui/main.ml]: Ps: i updated but it fails if the input code has backtick (`), will look into it later

**Comment by davesnx** [comment]
Thanks for the cc @Khady. It's much more functional with an editor for sure.

I would try to remove the script as JS and avoid the eval_string, but that's something we can fix in further PRs!

**Comment by Khady** [comment]
it works indeed and looks nice. There's a small problem when the code in the editor becomes too large. The editor grows and pushes the right column out of the screen

image

I played a bit with the lineWrapping option of codemirror, something like this should help

```diff
patdiff -git a/bin/web-ui/index.html b/bin/web-ui/index.html
No newline at the end of  a/bin/web-ui/index.html
index fbfdb0a..c37d4b4 100644
------  a/bin/web-ui/index.html
++++++  b/bin/web-ui/index.html
@|-25,32 +25,37 ============================================================
 |      width: 100%;
 |      display: flex;
 |      flex-direction: row;
 |    }
 |
 |    .flex-column {
 |      display: flex;
 |      flex-direction: column;
 |      flex-basis: 100%;
 |    }
 |
 |    @media (max-width: 800px) {
 |      .flex-row {
 |        flex-direction: column;
 |      }
 |    }
+|
+|    .CodeMirror-wrap {
+|      word-break: break-word;
+|    }
+|
 |  </style>
 |</head>
 |
 |<body>
 |  <div class="flex-column">
 |    <div class="flex-row">
 |      <div style="font-size: larger; font-weight: 700;">
 |        OCamlFormat configurator
 |      </div>
 |      <div style="margin-left: auto;">
 |        Made by <a href="https://ahrefs.com">Ahrefs</a> -
 |        Check our <a href="https://ahrefs.com/jobs">job openings</a> -
 |        Code available on <a href="https://github.com/ahrefs/ocamlformat/tree/ahrefs/web-ui/bin/web-ui">github</a>
 |      </div>
 |    </div>
 |    <div class="flex-row">
@|-66,25 +71,26 ============================================================
 |          <label for="config">OCamlFormat config
 |          </label>
 |          (<a
 |            href="https://raw.githubusercontent.com/ocaml-ppx/ocamlformat/main/ocamlformat-help.txt">documentation</a>)
 |        </div>
 |        <textarea name="config" id="config" rows="8" cols="80" placeholder="profile=default"
 |          spellcheck="false"></textarea>
 |        <div id="options"></div>
 |      </div>
 |    </div>
 |  </div>
 |  <script>
 |    window.editor = CodeMirror.fromTextArea(document.getElementById('code'), {
 |      mode: 'text/x-ocaml',
 |      smartIndent: true,
 |      lineNumbers: true,
+|      lineWrapping: true,
 |    });
 |
 |    editor.setSize('100%', '100%');
 |
 |    editor.save();
 |  </script>
 |</body>
 |
 |</html>
```

**Comment by pjmp** [comment]
[Applied your suggestion](https://github.com/ahrefs/ocamlformat/pull/12/commits/bc0251eabb7c3ae1df11916afb0f187827e9bab6) @Khady

File: bin/web-ui/index.html [hunk 3]
```diff
@@ -6,27 +6,42 @@
...
 
     @media (max-width: 800px) {
       .flex-row {
         flex-direction: column;
       }
     }
+
+    .CodeMirror-wrap {
+      word-break: break-word;
+    }
```
Comment (Khady) [inline bin/web-ui/index.html:44]: actually it looks like this part is not necessary (sorry I'm not a css expert ^^')

**Comment by Khady** [comment]
There's an issue with the alignment of the highlighting when some code is selected. I've no idea where it comes from. The problem doesn't seem to appear on the codemirror website

image
image

**Comment by Khady** [comment]
I'm tempted to merge this PR as it is (it's already better than the original version and it seems to work fine). Fixes/improvements could then be delivered through other PRs when people have time. This way you aren't rushed and everyone benefits from your work right now

**Comment by Khady** [comment]
also may I know why you decided to go with codemirror 5 instead of 6?

**Comment by pjmp** [comment]
I went for v5 instead of v6 purely out of familiarity and easy of use but I'm interested in porting this to v6 when i have time.

**Comment by Khady** [comment]
thanks a lot for the contribution!

**Pull request merged**